File: frontend/src/lib/schemas/baseProfile.ts
// Mirrors backend/schemas/base_profile.py. Keep in sync.
import { z } from "zod";

export const ymOrIsoDate = z
  .string()
  .regex(
    /^(\d{4}|\d{4}-\d{2}|\d{4}-\d{2}-\d{2})$/,
    "Expected YYYY, YYYY-MM, or YYYY-MM-DD",
  );

export const personalInfoSchema = z.object({
  full_name: z.string().min(1, "Full name is required"),
  email: z.string().email(),
  phone: z.string().optional(),
  location: z.string().optional(),
  headline: z.string().optional(),
  summary: z.string().optional(),
});

export const linkSchema = z.object({
  label: z.string().min(1),
  url: z.string().min(1),
});

export const experienceSchema = z.object({
  company: z.string().min(1, "Company is required"),
  title: z.string().min(1, "Title is required"),
  location: z.string().optional(),
  start_date: ymOrIsoDate,
  end_date: ymOrIsoDate.nullable().optional(),
  is_current: z.boolean(),
  bullets: z.array(z.string()),
  tech_stack: z.array(z.string()),
});

export const educationSchema = z.object({
  school: z.string().min(1, "School is required"),
  degree: z.string().min(1, "Degree is required"),
  field_of_study: z.string().optional(),
  location: z.string().optional(),
  start_date: ymOrIsoDate.optional(),
  end_date: ymOrIsoDate.optional(),
  is_current: z.boolean(),
  gpa: z.string().optional(),
  notes: z.string().optional(),
});

export const skillItemSchema = z.object({
  name: z.string().min(1),
  category: z.string().optional(),
  level: z.string().optional(),
  years: z.number().int().min(0).max(80).optional(),
});

export const projectSchema = z.object({
  name: z.string().min(1),
  role: z.string().optional(),
  description: z.string().optional(),
  bullets: z.array(z.string()),
  tech_stack: z.array(z.string()),
  url: z.string().optional(),
  start_date: ymOrIsoDate.optional(),
  end_date: ymOrIsoDate.optional(),
});

export const certificationSchema = z.object({
  name: z.string().min(1),
  issuer: z.string().optional(),
  issued_date: ymOrIsoDate.optional(),
  expiration_date: ymOrIsoDate.optional(),
  credential_id: z.string().optional(),
  url: z.string().optional(),
});

export const languageSchema = z.object({
  name: z.string().min(1),
  proficiency: z.string().optional(),
});

export const baseProfileSchema = z.object({
  personal: personalInfoSchema,
  links: z.array(linkSchema),
  experience: z.array(experienceSchema),
  education: z.array(educationSchema),
  skills: z.array(skillItemSchema),
  projects: z.array(projectSchema),
  certifications: z.array(certificationSchema),
  languages: z.array(languageSchema),
});

export const baseProfilePatchSchema = z.object({
  personal: personalInfoSchema.optional(),
  links: z.array(linkSchema).optional(),
  experience: z.array(experienceSchema).optional(),
  education: z.array(educationSchema).optional(),
  skills: z.array(skillItemSchema).optional(),
  projects: z.array(projectSchema).optional(),
  certifications: z.array(certificationSchema).optional(),
  languages: z.array(languageSchema).optional(),
});

export type PersonalInfo = z.infer<typeof personalInfoSchema>;
export type Link = z.infer<typeof linkSchema>;
export type Experience = z.infer<typeof experienceSchema>;
export type Education = z.infer<typeof educationSchema>;
export type SkillItem = z.infer<typeof skillItemSchema>;
export type Project = z.infer<typeof projectSchema>;
export type Certification = z.infer<typeof certificationSchema>;
export type Language = z.infer<typeof languageSchema>;
export type BaseProfile = z.infer<typeof baseProfileSchema>;
export type BaseProfilePatch = z.infer<typeof baseProfilePatchSchema>;

File: backend/schemas/base_profile.py
# backend/schemas/base_profile.py
"""Source-of-truth career profile.

Every tailored or skeleton output reads from this. Never invent fields
here — only the authenticated user writes to their own row.
"""
from typing import List, Optional
from pydantic import BaseModel, EmailStr, Field, field_validator


_DATE_RE_HINT = "Expected YYYY, YYYY-MM, or YYYY-MM-DD"


def _validate_date_loose(v: Optional[str]) -> Optional[str]:
    if v is None or v == "":
        return None
    if len(v) == 4 and v.isdigit():
        return v
    if len(v) in (7, 10) and v[4] == "-" and v[:4].isdigit() and v[5:7].isdigit():
        month = int(v[5:7])
        if not 1 <= month <= 12:
            raise ValueError(_DATE_RE_HINT)
        if len(v) == 10:
            if v[7] != "-" or not v[8:10].isdigit():
                raise ValueError(_DATE_RE_HINT)
            day = int(v[8:10])
            if not 1 <= day <= 31:
                raise ValueError(_DATE_RE_HINT)
        return v
    raise ValueError(_DATE_RE_HINT)


class PersonalInfo(BaseModel):
    full_name: str = Field(min_length=1)
    email: EmailStr
    phone: Optional[str] = None
    location: Optional[str] = None
    headline: Optional[str] = None
    summary: Optional[str] = None


class Link(BaseModel):
    label: str = Field(min_length=1)
    url: str = Field(min_length=1)


class Experience(BaseModel):
    company: str = Field(min_length=1)
    title: str = Field(min_length=1)
    location: Optional[str] = None
    start_date: str
    end_date: Optional[str] = None
    is_current: bool = False
    bullets: List[str] = []
    tech_stack: List[str] = []

    @field_validator("start_date")
    @classmethod
    def _start_date_format(cls, v: str) -> str:
        out = _validate_date_loose(v)
        if out is None:
            raise ValueError("start_date is required")
        return out

    @field_validator("end_date")
    @classmethod
    def _end_date_format(cls, v: Optional[str]) -> Optional[str]:
        return _validate_date_loose(v)


class Education(BaseModel):
    school: str = Field(min_length=1)
    degree: str = Field(min_length=1)
    field_of_study: Optional[str] = None
    location: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_current: bool = False
    gpa: Optional[str] = None
    notes: Optional[str] = None

    @field_validator("start_date", "end_date")
    @classmethod
    def _dates(cls, v: Optional[str]) -> Optional[str]:
        return _validate_date_loose(v)


class SkillItem(BaseModel):
    name: str = Field(min_length=1)
    category: Optional[str] = None
    level: Optional[str] = None
    years: Optional[int] = Field(default=None, ge=0, le=80)


class Project(BaseModel):
    name: str = Field(min_length=1)
    role: Optional[str] = None
    description: Optional[str] = None
    bullets: List[str] = []
    tech_stack: List[str] = []
    url: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None

    @field_validator("start_date", "end_date")
    @classmethod
    def _dates(cls, v: Optional[str]) -> Optional[str]:
        return _validate_date_loose(v)


class Certification(BaseModel):
    name: str = Field(min_length=1)
    issuer: Optional[str] = None
    issued_date: Optional[str] = None
    expiration_date: Optional[str] = None
    credential_id: Optional[str] = None
    url: Optional[str] = None

    @field_validator("issued_date", "expiration_date")
    @classmethod
    def _dates(cls, v: Optional[str]) -> Optional[str]:
        return _validate_date_loose(v)


class Language(BaseModel):
    name: str = Field(min_length=1)
    proficiency: Optional[str] = None


class BaseProfile(BaseModel):
    personal: PersonalInfo
    links: List[Link] = []
    experience: List[Experience] = []
    education: List[Education] = []
    skills: List[SkillItem] = []
    projects: List[Project] = []
    certifications: List[Certification] = []
    languages: List[Language] = []


class BaseProfilePatch(BaseModel):
    """JSON-Merge-Patch style partial update. Any present key replaces the
    full corresponding key on the stored profile; absent keys are untouched."""
    personal: Optional[PersonalInfo] = None
    links: Optional[List[Link]] = None
    experience: Optional[List[Experience]] = None
    education: Optional[List[Education]] = None
    skills: Optional[List[SkillItem]] = None
    projects: Optional[List[Project]] = None
    certifications: Optional[List[Certification]] = None
    languages: Optional[List[Language]] = None
